File: backend/app/services/website.py
from urllib.parse import urlparse

from app.schemas.security import (
    CertificateInfo,
    OpenPort,
    SslInfo,
    Vulnerability,
    WebsiteScanResult,
)


def scan_website(url: str) -> WebsiteScanResult:
    raw = url.strip()
    if "://" not in raw:
        raw = f"https://{raw}"

    try:
        hostname = urlparse(raw).hostname or raw
    except Exception:
        hostname = raw.replace("https://", "").replace("http://", "").split("/")[0]

    https_status = raw.startswith("https://")
    domain_reputation = 76 if https_status else 34
    overall_score = 72 if https_status else 38

    if https_status:
        vulns = [
            Vulnerability(
                cve="CVE-2024-XXXX",
                severity="medium",
                description="Outdated server banner may disclose version information.",
            )
        ]
        recommendations = [
            "Enable HSTS to enforce HTTPS for all visitors",
            "Implement Content-Security-Policy headers",
            "Disable directory listing and verbose error pages",
            "Keep server software patched against latest CVEs",
            "Restrict management ports (SSH/RDP) to VPN or allowlists",
            "Enable WAF rules for common OWASP Top 10 attacks",
        ]
    else:
        vulns = [
            Vulnerability(
                cve="N/A-HTTPS-MISSING",
                severity="critical",
                description="Site does not enforce HTTPS — traffic can be intercepted.",
            ),
            Vulnerability(
                cve="CVE-2025-19882",
                severity="high",
                description="Potential path traversal if running vulnerable Apache versions.",
            ),
        ]
        recommendations = [
            "Immediately enable HTTPS with a valid certificate",
            "Implement Content-Security-Policy headers",
            "Disable directory listing and verbose error pages",
            "Keep server software patched against latest CVEs",
            "Restrict management ports (SSH/RDP) to VPN or allowlists",
            "Enable WAF rules for common OWASP Top 10 attacks",
        ]

    return WebsiteScanResult(
        url=raw,
        httpsStatus=https_status,
        sslInfo=SslInfo(
            issuer="Let's Encrypt Authority X3" if https_status else "None",
            validFrom="2025-03-01",
            validTo="2026-03-01",
            grade="A-" if https_status else "F",
            protocol="TLS 1.3" if https_status else "N/A",
        ),
        domainReputation=domain_reputation,
        certificateInfo=CertificateInfo(
            subject=f"CN={hostname}",
            san=[f"DNS:{hostname}", f"DNS:www.{hostname}"],
            signatureAlgorithm="SHA256-RSA",
        ),
        openPorts=[
            OpenPort(port=80, service="HTTP", status="open"),
            OpenPort(port=443, service="HTTPS", status="open" if https_status else "closed"),
            OpenPort(port=22, service="SSH", status="filtered"),
            OpenPort(port=3389, service="RDP", status="closed"),
        ],
        knownVulnerabilities=vulns,
        securityRecommendations=recommendations,
        overallScore=overall_score,
    )

File: lib/server/website.ts
import type { WebsiteScanResult } from "@/types";

export function scanWebsite(url: string): WebsiteScanResult {
  let raw = url.trim();
  if (!raw.includes("://")) raw = `https://${raw}`;

  let hostname = raw;
  try {
    hostname = new URL(raw).hostname;
  } catch {
    hostname = raw.replace(/^https?:\/\//, "").split("/")[0];
  }

  const httpsStatus = raw.startsWith("https://");
  const domainReputation = httpsStatus ? 76 : 34;
  const overallScore = httpsStatus ? 72 : 38;

  return {
    url: raw,
    httpsStatus,
    sslInfo: {
      issuer: httpsStatus ? "Let's Encrypt Authority X3" : "None",
      validFrom: "2025-03-01",
      validTo: "2026-03-01",
      grade: httpsStatus ? "A-" : "F",
      protocol: httpsStatus ? "TLS 1.3" : "N/A",
    },
    domainReputation,
    certificateInfo: {
      subject: `CN=${hostname}`,
      san: [`DNS:${hostname}`, `DNS:www.${hostname}`],
      signatureAlgorithm: "SHA256-RSA",
    },
    openPorts: [
      { port: 80, service: "HTTP", status: "open" },
      { port: 443, service: "HTTPS", status: httpsStatus ? "open" : "closed" },
      { port: 22, service: "SSH", status: "filtered" },
      { port: 3389, service: "RDP", status: "closed" },
    ],
    knownVulnerabilities: httpsStatus
      ? [
          {
            cve: "CVE-2024-XXXX",
            severity: "medium",
            description: "Outdated server banner may disclose version information.",
          },
        ]
      : [
          {
            cve: "N/A-HTTPS-MISSING",
            severity: "critical",
            description: "Site does not enforce HTTPS — traffic can be intercepted.",
          },
          {
            cve: "CVE-2025-19882",
            severity: "high",
            description: "Potential path traversal if running vulnerable Apache versions.",
          },
        ],
    securityRecommendations: [
      httpsStatus
        ? "Enable HSTS to enforce HTTPS for all visitors"
        : "Immediately enable HTTPS with a valid certificate",
      "Implement Content-Security-Policy headers",
      "Disable directory listing and verbose error pages",
      "Keep server software patched against latest CVEs",
      "Restrict management ports (SSH/RDP) to VPN or allowlists",
      "Enable WAF rules for common OWASP Top 10 attacks",
    ],
    overallScore,
  };
}
